Investigate using simd-json for performance improvements

In production :hotdog: is typically CPU bound on JSON parsing, JSON serialization, and Handlebars template rendering.

[simd-json](https://github.com/simd-lite/simd-json) might be a good route to explore for enabling for platforms that support SIMD.

At first glance it looks like simd-json is not API compatible with serde, since its `from_str` and `from_slice` methods will actually consume the buffer :facepalm: 

```
   Compiling hotdog v0.3.3 (/home/tyler/source/github/reiseburo/hotdog)
error[E0308]: mismatched types
  --> src/json.rs:15:41
   |
15 |         simd_json::serde::from_str::<S>(buffer)
   |                                         ^^^^^^ types differ in mutability
   |
   = note: expected mutable reference `&mut str`
                      found reference `&'a str`

error[E0308]: mismatched types
  --> src/json.rs:15:9
   |
12 | pub fn from_str<'a, S: serde::Deserialize<'a>>(buffer: &'a str) -> Result<S, serde_json::error::Error> {
   |                                                                    ----------------------------------- expected `std::result::Result<S, serde_json::error::Error>` because of return type
...
15 |         simd_json::serde::from_str::<S>(buffer)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `serde_json::error::Error`, found struct `simd_json::Error`
   |
   = note: expected enum `std::result::Result<_, serde_json::error::Error>`
              found enum `std::result::Result<_, simd_json::Error>`

error: aborting due to 2 previous errors
```

--- a/.cargo/config
+++ b/.cargo/config
@@ -1,0 +1,2 @@
+[build]
+rustflags = ["-C", "target-cpu=native"]

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -27,3 +27,5 @@ jobs:
       run: cargo build --verbose
     - name: Run tests
       run: cargo test --verbose
+    - name: Run tests with simd
+      run: cargo test --verbose --features simd

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -4,7 +4,10 @@ version = "0.3.3"
 authors = ["R. Tyler Croy <rtyler+hotdog@brokenco.de>"]
 edition = "2018"
 
-# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[features]
+default = []
+simd = ["simd-json"]
 
 [dependencies]
 # Base asynchronous runtime
@@ -52,6 +55,9 @@ serde_json = "~1.0.0"
 # Helpful to deserialize our regular expressions directly from the
 # configuration file
 serde_regex = "~0.4.0"
+
+simd-json = { version = "~0.3.14", optional = true}
+
 smol = "~0.1.18"
 # Needed to do clever enum/derive tricks for strings
 strum = "0.18.0"

--- a/src/connection.rs
+++ b/src/connection.rs
@@ -317,9 +317,9 @@ fn precompile_jmespath(map: &mut JmesPathExpressions, settings: Arc<Settings>) -
  * perform_merge will generate the buffer resulting of the JSON merge
  */
 fn perform_merge(buffer: &str, template_id: &str, state: &RuleState) -> Result<String, String> {
-    if let Ok(mut msg_json) = serde_json::from_str(&buffer) {
+    if let Ok(mut msg_json) = crate::json::from_str(&buffer) {
         if let Ok(rendered) = state.hb.render(template_id, &state.variables) {
-            let to_merge: serde_json::Value = serde_json::from_str(&rendered)
+            let to_merge: serde_json::Value = crate::json::from_str(&rendered)
                 .expect("Failed to deserialize our rendered to_merge_str");
 
             /*
@@ -333,7 +333,7 @@ fn perform_merge(buffer: &str, template_id: &str, state: &RuleState) -> Result<S
 
             merge::merge(&mut msg_json, &to_merge);
 
-            if let Ok(output) = serde_json::to_string(&msg_json) {
+            if let Ok(output) = crate::json::to_string(&msg_json) {
                 return Ok(output);
             }
         }

--- a/src/connection.rs
+++ b/src/connection.rs
@@ -99,8 +99,11 @@ impl Connection {
             }
             /*
              * Now that we've logged the error, let's unpack and bubble the error anyways
+             *
+             * Note: msg needs to be mutable so we can fish the `msg` out within it during a
+             * simd_json parse
              */
-            let msg = parsed.unwrap();
+            let mut msg = parsed.unwrap();
             self.stats.send((Stats::LineReceived, 1)).await;
             let mut continue_rules = true;
             debug!("parsed as: {}", msg.msg);
@@ -215,9 +218,11 @@ impl Connection {
 
                         Action::Merge { json, json_str: _ } => {
                             debug!("merging JSON content: {}", json);
-                            if let Ok(buffer) =
-                                perform_merge(&msg.msg, &template_id_for(&rule, index), &rule_state)
-                            {
+                            if let Ok(buffer) = perform_merge(
+                                &mut msg.msg,
+                                &template_id_for(&rule, index),
+                                &rule_state,
+                            ) {
                                 output = buffer;
                             } else {
                                 continue_rules = false;
@@ -316,10 +321,14 @@ fn precompile_jmespath(map: &mut JmesPathExpressions, settings: Arc<Settings>) -
 /**
  * perform_merge will generate the buffer resulting of the JSON merge
  */
-fn perform_merge(buffer: &str, template_id: &str, state: &RuleState) -> Result<String, String> {
-    if let Ok(mut msg_json) = crate::json::from_str(&buffer) {
-        if let Ok(rendered) = state.hb.render(template_id, &state.variables) {
-            let to_merge: serde_json::Value = crate::json::from_str(&rendered)
+fn perform_merge(
+    mut buffer: &mut str,
+    template_id: &str,
+    state: &RuleState,
+) -> Result<String, String> {
+    if let Ok(mut msg_json) = crate::json::from_str(&mut buffer) {
+        if let Ok(mut rendered) = state.hb.render(template_id, &state.variables) {
+            let to_merge: serde_json::Value = crate::json::from_str(&mut rendered)
                 .expect("Failed to deserialize our rendered to_merge_str");
 
             /*
@@ -374,7 +383,8 @@ mod tests {
         let hash = HashMap::<String, String>::new();
         let state = rule_state(&hb, &hash);
 
-        let output = perform_merge("{}", template_id, &state);
+        let mut buffer = "{}".to_string();
+        let output = perform_merge(&mut buffer, template_id, &state);
         assert_eq!(output, Ok("{}".to_string()));
     }
 
@@ -390,7 +400,8 @@ mod tests {
         let hash = HashMap::<String, String>::new();
         let state = rule_state(&hb, &hash);
 
-        let output = perform_merge("{}", template_id, &state)?;
+        let mut buffer = "{}".to_string();
+        let output = perform_merge(&mut buffer, template_id, &state)?;
         assert_eq!(output, "{}".to_string());
         Ok(())
     }
@@ -407,7 +418,8 @@ mod tests {
         let hash = HashMap::<String, String>::new();
         let state = rule_state(&hb, &hash);
 
-        let output = perform_merge("invalid", template_id, &state);
+        let mut buffer = "invalid".to_string();
+        let output = perform_merge(&mut buffer, template_id, &state);
         let expected = Err("Not JSON".to_string());
         assert_eq!(output, expected);
     }
@@ -424,7 +436,8 @@ mod tests {
         let hash = HashMap::<String, String>::new();
         let state = rule_state(&hb, &hash);
 
-        let output = perform_merge("{}", template_id, &state);
+        let mut buffer = "{}".to_string();
+        let output = perform_merge(&mut buffer, template_id, &state);
         assert_eq!(output, Ok("{\"hello\":1}".to_string()));
     }
 
@@ -441,7 +454,8 @@ mod tests {
         hash.insert("name".to_string(), "world".to_string());
         let state = rule_state(&hb, &hash);
 
-        let output = perform_merge("{}", template_id, &state);
+        let mut buffer = "{}".to_string();
+        let output = perform_merge(&mut buffer, template_id, &state);
         assert_eq!(output, Ok("{\"hello\":\"world\"}".to_string()));
     }
 

--- a/src/json.rs
+++ b/src/json.rs
@@ -1,0 +1,26 @@
+/**
+ * This module acts as a shim between serde_json and simd-json to allow for higher performance JSON
+ * parsing on SIMD-capable architectures
+ */
+
+
+use serde;
+use serde_json;
+#[cfg(feature = "simd")]
+use simd_json;
+
+pub fn from_str<'a, S: serde::Deserialize<'a>>(buffer: &'a str) -> Result<S, serde_json::error::Error> {
+    #[cfg(feature = "simd")]
+    {
+        simd_json::serde::from_str::<S>(buffer)
+    }
+
+    #[cfg(not(feature="simd"))]
+    {
+        serde_json::from_str::<S>(buffer)
+    }
+}
+
+pub fn to_string<S: serde::Serialize>(value: &S) -> Result<String, serde_json::error::Error> {
+    serde_json::to_string::<S>(value)
+}

--- a/src/json.rs
+++ b/src/json.rs
@@ -2,22 +2,21 @@
  * This module acts as a shim between serde_json and simd-json to allow for higher performance JSON
  * parsing on SIMD-capable architectures
  */
-
-
 use serde;
 use serde_json;
 #[cfg(feature = "simd")]
 use simd_json;
 
-pub fn from_str<'a, S: serde::Deserialize<'a>>(buffer: &'a str) -> Result<S, serde_json::error::Error> {
+pub fn from_str<'a, S: serde::Deserialize<'a>>(buffer: &'a mut str) -> Result<S, std::io::Error> {
     #[cfg(feature = "simd")]
     {
         simd_json::serde::from_str::<S>(buffer)
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))
     }
 
-    #[cfg(not(feature="simd"))]
+    #[cfg(not(feature = "simd"))]
     {
-        serde_json::from_str::<S>(buffer)
+        serde_json::from_str::<S>(buffer).map_err(|e| e.into())
     }
 }
 

--- a/src/main.rs
+++ b/src/main.rs
@@ -13,6 +13,8 @@ extern crate serde;
 #[macro_use]
 extern crate serde_derive;
 extern crate serde_regex;
+#[cfg(feature = "simd")]
+extern crate simd_json;
 extern crate strum;
 extern crate syslog_loose;
 extern crate syslog_rfc5424;

--- a/src/main.rs
+++ b/src/main.rs
@@ -28,6 +28,7 @@ use log::*;
 
 mod connection;
 mod errors;
+mod json;
 mod kafka;
 mod merge;
 mod parse;

